### DISCUSSION TITLE
Usage with Remix and Playwright

### DISCUSSION THREAD

**Discussion by kirillkurko** [mswjs/msw#1668] [Q&A] [answered]
I'm looking for a good example of using MSW with Remix and Playwright. There are some examples on the internet like [this one](https://github.com/cliffordfajardo/remix-msw-node-with-playwright), but what I'm looking for is the usage of `server.use(...)` inside my tests. The problem that I'm facing is that while handlers are prepended (verified using `server.printHandlers()`), override is not working.

**Comment by kettanaito** [comment, marked as answer]
Hey, @kirillkurko. Take a look at our [Official Remix + Playwright + MSW](https://github.com/mswjs/examples/tree/main/examples/with-remix) usage example. 

Using the runtime handlers (`worker.use()`) is a matter of exposing `worker` on `window` and accessing it in `page.evaluate()` in your Playwright tests.

**Reply by kirillkurko** [reply]
Thanks for the reply! So the `worker` is the only way to do it? I don't use it currently and would like to avoid adding it because of the e2e tests, if possible

**Reply by kettanaito** [reply]
Can you tell me more about what you're trying to do?

Using Playwright implies running in the browser, and the way to use MSW in the browser is via the [setupWorker](https://mswjs.io/docs/api/setup-worker) API. I may be missing the point of what you're trying to achieve, would like to learn more.

**Reply by kettanaito** [reply]
Note that the example I linked features both client-side and server-side API mocking in Remix at the same time. You cannot use `server.use()` in Playwright tests because your app and your test are running in two separate processes. I'm playing around with a solution for this in #1617, which is going to be huge. Until then, you cannot have test-based server-side API mocks with MSW, just as you can't have them in general at the moment.

**Reply by kirillkurko** [reply]
Thanks for the reply, you actually answered my question and `worker` is a way to go for me. "setupRemoteServer" API looks really interesting, but it's definitely a huge milestone

**Reply by pauloricardos** [reply]
hey @kettanaito this page [Official Remix + Playwright + MSW](https://github.com/mswjs/examples-new/tree/main/examples/with-remix) seems broken

**Reply by kettanaito** [reply]
@pauloricardos, thanks for letting me know! So GitHub repo redirects do have an EOL. Fixed the link, it leads to the correct example now.

**Comment by dadamssg** [comment]
I know this is an old post but i came up with something that works. I created a remix action to interact with `server.use()` from my playwright tests. 

<img width="1246" alt="image" src="https://github.com/user-attachments/assets/7caa4ba8-4750-411c-9cf6-d452f09b97ac">

Super low tech solution to the remote server concept. This might be extremely dumb though 🤷‍♂️😅

**Reply by jumpcutfindo** [reply]
Hi, thanks for the comment, we were facing something similar in our project and this led us down a similar path :)

We implemented an action similar to yours, but to add more fine grained control we implemented a simple API to control the mock server:

```typescript
// mocks/mock-controller.ts

export async function action({ request }: ActionFunctionArgs) {
  // Parse the request here and switch according to some fields in the request body
  // Depending on the action specified, route to the appropriate functions
}

async function startup() { ... }
async function shutdown() { ... }
async function reset() { ... }
async function override(request: OverrideRequest) { ... }
```

To make it more elegant inside each Playwright test, we added some [Playwright fixtures](https://playwright.dev/docs/test-fixtures), inspired in part by the implementation in [playwright-msw](https://github.com/valendres/playwright-msw):
```typescript

// tests/fixtures.ts

interface TestFixtures {
  mockServer: {
    startup: () => Promise<void>;
    shutdown: () => Promise<void>;
    reset: () => Promise<void>;
    override: (data: OverrideRequestBody) => Promise<void>;
  };
}

const test = base.extend<TestFixtures>({
  mockServer: {
    startup: async () => {
      await fetch(MOCK_CONTROLLER_ENDPOINT, {
        method: 'POST',
        body: JSON.stringify({
          action: 'startup',
        }),
      });
    },

    shutdown: async () => {
      await fetch(MOCK_CONTROLLER_ENDPOINT, {
        method: 'POST',
        body: JSON.stringify({
          action: 'shutdown',
        }),
      });
    },

    reset: async () => {
      await fetch(MOCK_CONTROLLER_ENDPOINT, {
        method: 'POST',
        body: JSON.stringify({
          action: 'reset',
        }),
      });
    },

    override: async (data) => {
      await fetch(MOCK_CONTROLLER_ENDPOINT, {
        method: 'POST',
        body: JSON.stringify({
          action: 'override',
          data,
        }),
      });
    },
  },
});

export { test, expect };
```

This allows us to configure the mock server response like so in individual tests:

``` typescript
// tests/login.spec.ts

// Note:
// - Need to start, reset and stop mock server with `test.beforeAll`, `test.afterEach` and `test.afterAll` with this
// - Probably have to run the test suite with `test.describe.serial` instead of in parallel

import { expect, test } from 'tests/fixtures';

test('should show error when trying invalid credentials', async ({
    page,
    mockServer,
  }) => {
    await mockServer.override({
      method: HttpMethods.POST,
      url: '/user/login',
      statusCode: 401,
      body: { error: 'Incorrect username or password' },
    });
   
   ... rest of the test
 });
```

We also configured it such that the mock controller is registered only if process.env.NODE_ENV is test:
```typescript
// vite.config.ts

export default defineConfig({
  plugins: [
    ... other plugins,
    routes(defineRoutes) {
        return defineRoutes((route) => {
          if (process.env.NODE_ENV === 'test') {
            // Register mocked route for handling
            route('/__mock', '../mocks/mock-controller.ts');
          }
        });
      },
   ],
});
```

**Reply by dadamssg** [reply]
nice!

**Reply by AriPerkkio** [reply]
@dadamssg & @jumpcutfindo, are you running these setups against the Vite dev server or Remix production build?
